> Introdução à Ciência dos Dados \
> Ciência da Computação – UFV/Florestal


* **Nome:** %s
* **Matrícula:**  %d

# Aula Prática 07
## Regras de Associação e o Algoritmo Apriori

**Objetivo:** Aplicar o algoritmo Apriori para extrair regras de associação de dados reais.
**Pré-requisitos:** Linguagem de programação Python, estatística básica \
**Meta:** Ao final da prática, o aluno será capaz de preparar dados e extrair regras de associação desses dados utilizado o algoritmo Apriori.

----

In [ ]:
### NÃO REMOVA ESSA CÉLULA! ####
!pip --no-cache-dir install --upgrade learntools_dados_ufv

!wget https://raw.githubusercontent.com/NESPEDUFV/CCF425-resources/master/p7/Market_Basket_Optimisation.csv

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
--2023-05-24 15:16:41--  https://raw.githubusercontent.com/NESPEDUFV/CCF425-resources/master/p7/Market_Basket_Optimisation.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 302908 (296K) [text/plain]
Saving to: ‘Market_Basket_Optimisation.csv.2’

Market_Basket_Optim 100%[===================>] 295.81K  --.-KB/s    in 0.03s   

2023-05-24 15:16:41 (9.21 MB/s) - ‘Market_Basket_Optimisation.csv.2’ saved [302908/302908]



In [ ]:
#### Não altere essa célula!!! ####
import pandas as pd
import string
import random
import numpy as np

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

from learntools_dados_ufv.core import binder; binder.bind(globals())
from learntools_dados_ufv.ccf425.pratica_7b import *

# Roteiro

## Estruturando os dados

### Criar o conjunto de dados

In [ ]:
dataset = [['Leite', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
          ['Arroz', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
          ['Leite', 'Maça', 'Feijão', 'Ovo'],
          ['Leite', 'Chocolate', 'Pipoca', 'Feijão', 'Iogurte'],
          ['Pipoca', 'Cebola', 'Feijão', 'Sorvete', 'Ovo']]

dataset

[['Leite', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
 ['Arroz', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
 ['Leite', 'Maça', 'Feijão', 'Ovo'],
 ['Leite', 'Chocolate', 'Pipoca', 'Feijão', 'Iogurte'],
 ['Pipoca', 'Cebola', 'Feijão', 'Sorvete', 'Ovo']]

### Transformar no formato para ser utilizado no Apriori

In [ ]:
transactions_oht = TransactionEncoder()
transactions = transactions_oht.fit(dataset).transform(dataset)
transactions

array([[False,  True, False,  True,  True,  True, False,  True,  True,
        False, False],
       [ True,  True, False,  True,  True, False, False,  True,  True,
        False, False],
       [False, False, False,  True, False,  True,  True, False,  True,
        False, False],
       [False, False,  True,  True,  True,  True, False, False, False,
         True, False],
       [False,  True, False,  True, False, False, False, False,  True,
         True,  True]])

In [ ]:
df = pd.DataFrame(transactions, columns=transactions_oht.columns_)
df

Arroz  Cebola  Chocolate  Feijão  Iogurte  Leite   Maça  Nutela    Ovo  \
0  False    True      False    True     True   True  False    True   True   
1   True    True      False    True     True  False  False    True   True   
2  False   False      False    True    False   True   True   False   True   
3  False   False       True    True     True   True  False   False  False   
4  False    True      False    True    False  False  False   False   True   

   Pipoca  Sorvete  
0   False    False  
1   False    False  
2   False    False  
3    True    False  
4    True     True

## Calcular conjunto de itens frequentes

Com suporte mínimo de 0.6

In [ ]:
# Calcula os conjuntos de itens frequentes, com suporte minimo de 0.6
frequent_itemsets = apriori(df, min_support=0.6, use_colnames=True)
frequent_itemsets

support               itemsets
0       0.6               (Cebola)
1       1.0               (Feijão)
2       0.6              (Iogurte)
3       0.6                (Leite)
4       0.8                  (Ovo)
5       0.6       (Cebola, Feijão)
6       0.6          (Ovo, Cebola)
7       0.6      (Iogurte, Feijão)
8       0.6        (Feijão, Leite)
9       0.8          (Ovo, Feijão)
10      0.6  (Ovo, Cebola, Feijão)

## Calcular regras de associacão utilizando o Apriori

Com filtro de confiança de 0.1

In [ ]:
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.1)
rules

antecedents       consequents  antecedent support  \
0           (Cebola)          (Feijão)                 0.6   
1           (Feijão)          (Cebola)                 1.0   
2              (Ovo)          (Cebola)                 0.8   
3           (Cebola)             (Ovo)                 0.6   
4          (Iogurte)          (Feijão)                 0.6   
5           (Feijão)         (Iogurte)                 1.0   
6           (Feijão)           (Leite)                 1.0   
7            (Leite)          (Feijão)                 0.6   
8              (Ovo)          (Feijão)                 0.8   
9           (Feijão)             (Ovo)                 1.0   
10     (Ovo, Cebola)          (Feijão)                 0.6   
11     (Ovo, Feijão)          (Cebola)                 0.8   
12  (Cebola, Feijão)             (Ovo)                 0.6   
13             (Ovo)  (Cebola, Feijão)                 0.8   
14          (Cebola)     (Ovo, Feijão)                 0.6   
15          (Feijão)     (Ovo, Cebola)                 1.0   

    consequent support  support  confidence  lift  leverage  conviction  
0                  1.0      0.6        1.00  1.00      0.00         inf  
1                  0.6      0.6        0.60  1.00      0.00         1.0  
2                  0.6      0.6        0.75  1.25      0.12         1.6  
3                  0.8      0.6        1.00  1.25      0.12         inf  
4                  1.0      0.6        1.00  1.00      0.00         inf  
5                  0.6      0.6        0.60  1.00      0.00         1.0  
6                  0.6      0.6        0.60  1.00      0.00         1.0  
7                  1.0      0.6        1.00  1.00      0.00         inf  
8                  1.0      0.8        1.00  1.00      0.00         inf  
9                  0.8      0.8        0.80  1.00      0.00         1.0  
10                 1.0      0.6        1.00  1.00      0.00         inf  
11                 0.6      0.6        0.75  1.25      0.12         1.6  
12                 0.8      0.6        1.00  1.25      0.12         inf  
13                 0.6      0.6        0.75  1.25      0.12         1.6  
14                 0.8      0.6        1.00  1.25      0.12         inf  
15                 0.6      0.6        0.60  1.00      0.00         1.0

In [ ]:
# Filtro por 'confidence'. Pode ser feito por outros campos também
rules[rules['confidence'] > 0.75]

antecedents    consequents  antecedent support  consequent support  \
0           (Cebola)       (Feijão)                 0.6                 1.0   
3           (Cebola)          (Ovo)                 0.6                 0.8   
4          (Iogurte)       (Feijão)                 0.6                 1.0   
7            (Leite)       (Feijão)                 0.6                 1.0   
8              (Ovo)       (Feijão)                 0.8                 1.0   
9           (Feijão)          (Ovo)                 1.0                 0.8   
10     (Ovo, Cebola)       (Feijão)                 0.6                 1.0   
12  (Cebola, Feijão)          (Ovo)                 0.6                 0.8   
14          (Cebola)  (Ovo, Feijão)                 0.6                 0.8   

    support  confidence  lift  leverage  conviction  
0       0.6         1.0  1.00      0.00         inf  
3       0.6         1.0  1.25      0.12         inf  
4       0.6         1.0  1.00      0.00         inf  
7       0.6         1.0  1.00      0.00         inf  
8       0.8         1.0  1.00      0.00         inf  
9       0.8         0.8  1.00      0.00         1.0  
10      0.6         1.0  1.00      0.00         inf  
12      0.6         1.0  1.25      0.12         inf  
14      0.6         1.0  1.25      0.12         inf

# Exercícios

Os exercícios abaixo são referentes ao conjunto de dados `Market_Basket_Optimisation.csv`. Esse conjunto possui compras considerando 20 produtos distintos.

Execute e não remova a célula abaixo antes de iniciar os exercícios. Faça todas as atividades manipulando a variável `df`.

In [ ]:
# Atribui o dataframe com o estado "original" em uma variável isolada
original_df = pd.read_csv('./Market_Basket_Optimisation.csv', header=None)
df = original_df.copy()
df

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs               NaN   
2           chutney                NaN          NaN               NaN   
3            turkey            avocado          NaN               NaN   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread               NaN   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken                NaN          NaN               NaN   
7499       escalope          green tea          NaN               NaN   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1              NaN               NaN   NaN             NaN           NaN   
2              NaN               NaN   NaN             NaN           NaN   
3              NaN               NaN   NaN             NaN           NaN   
4        green tea               NaN   NaN             NaN           NaN   
...            ...               ...   ...             ...           ...   
7496           NaN               NaN   NaN             NaN           NaN   
7497     magazines         green tea   NaN             NaN           NaN   
7498           NaN               NaN   NaN             NaN           NaN   
7499           NaN               NaN   NaN             NaN           NaN   
7500           NaN               NaN   NaN             NaN           NaN   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1              NaN             NaN        NaN    NaN    NaN            NaN   
2              NaN             NaN        NaN    NaN    NaN            NaN   
3              NaN             NaN        NaN    NaN    NaN            NaN   
4              NaN             NaN        NaN    NaN    NaN            NaN   
...            ...             ...        ...    ...    ...            ...   
7496           NaN             NaN        NaN    NaN    NaN            NaN   
7497           NaN             NaN        NaN    NaN    NaN            NaN   
7498           NaN             NaN        NaN    NaN    NaN            NaN   
7499           NaN             NaN        NaN    NaN    NaN            NaN   
7500           NaN             NaN        NaN    NaN    NaN            NaN   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1        NaN                NaN              NaN      NaN        NaN  
2        NaN                NaN              NaN      NaN        NaN  
3        NaN                NaN              NaN      NaN        NaN  
4        NaN                NaN              NaN      NaN        NaN  
...      ...                ...              ...      ...        ...  
7496     NaN                NaN              NaN      NaN        NaN  
7497     NaN                NaN              NaN      NaN        NaN  
7498     NaN                NaN              NaN      NaN        NaN  
7499     NaN                NaN              NaN      NaN        NaN  
7500     NaN                NaN              NaN      NaN        NaN  

[7501 rows x 20 columns]

In [ ]:
len(df.columns)

20

## 1. Faça uma preparação dos dados.
Antes de fazer a análise de regras de associação, é necessário que você trate alguns dados desse dataset. Essa prática é **extremamente** importante e comum, já que os dados que encontramos ~~quase nunca~~ nem sempre estão prontos para serem manipulados.

### 1.a. Crie um dataframe com base nas transações, usando o TransactionEncoder e manipulação de Array

Chame a função `q1.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q1.a.hint()`

In [ ]:
# Resposta da 1

# q1.hint()
transactions = []
for i in range(0, 7501):
    aux = []
    for j in range(0, 20):
      if str(df.values[i,j]) != 'nan': aux.append(df.values[i,j])
    transactions.append(aux)

transactions_oht = TransactionEncoder()
transactions = transactions_oht.fit(transactions).transform(transactions)
df = pd.DataFrame(transactions, columns=transactions_oht.columns_)
q1.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



## 2. Faça o cálculo dos itemsets frequentes e regras de associação.
Utilize o algoritmo _apriori_.

### 2.a. A partir de `df`, calcule os itemsets frequentes utilizando `apriori`. 

Utilize o suporte mínimo de 0.01. Utilize o paramêtro `use_colnames=True`.

Armazene-a na variável `frequent_itemsets`.

Chame a função `q3.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.a.hint()`

In [ ]:
# Resposta da 2.a

# q2.a.hint()

frequent_itemsets = apriori(df, min_support=0.01, use_colnames=True)

q2.a.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



### 2.b. A partir de `frequent_itemsets`, calcule as regras de associação. 

Utilize filtro de confiança de `0.3`.

Armazene-a na variável `rules`.

Chame a função `q2.b.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q2.b.hint()`

In [ ]:
# Resposta da 2.b

# q2.b.hint()

rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.3)
q2.b.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



## 3. Analise as regras obtidas.

### 3.a. A partir de `rules`, recupere as regras com suporte maior ou igual a 0.015 e confiança maior ou igual a 0.4. 

Armazene-a na variável `rules_a`.

Chame a função `q3.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.a.hint()`

In [ ]:
# Resposta da 3.a

# q3.a.hint()

rules_a = rules[(rules['support']>=0.015) & (rules['confidence']>=0.4)]

q3.a.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



### 3.b. A partir de `rules`, recupere as regras com lift maior ou igual a 2 e confiança maior ou igual a 0.5. 

Armazene-a na variável `rules_b`.

Chame a função `q3.b.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.b.hint()`

In [ ]:
# Resposta da 3.b

# q3.b.hint()

rules_b = rules[(rules['lift']>=2) & (rules['confidence']>=0.5)]

q3.b.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



### 3.c. Quais produtos você recomendaria colocar próximos se fosse organizar um supermercado físico?

Justifique sua resposta com evidências plausíveis em uma resposta completa.

Essa questão não possui check automático. Mas você pode pedir dica com `q3.c.hint()`. 

**RESPOSTA 3.c.:** 

Existem vários fatores que podem influenciar na organização dos produtos no supermecado. Além das associações usando frequência que os itens aparecem juntos, confiança e lift, há também temáticas, fluxos e etc.

Mas olhando apenas para os dados, uma forma de descobrir alguns dos melhores produtos para serem colocados próximos seria as associações de suporte e confiança.

Com base nisso, estabelecemos os itens que contêm respectivamente os melhores valores de confiança e suporte.

Olhando para a tabela abaixo, encontramos algumas combinações que são até mesmo tradições culinárias como Vinho e Massa, no caso, (red wine, spaghetti)



In [ ]:
rules.sort_values(['support','confidence'], ascending=[False, False])

antecedents      consequents  antecedent support  \
29                    (spaghetti)  (mineral water)            0.174110   
5                     (chocolate)  (mineral water)            0.163845   
21                         (milk)  (mineral water)            0.129583   
15                  (ground beef)  (mineral water)            0.098254   
16                  (ground beef)      (spaghetti)            0.098254   
..                            ...              ...                 ...   
60     (mineral water, olive oil)      (spaghetti)            0.027596   
34                     (red wine)      (spaghetti)            0.028130   
45            (eggs, ground beef)  (mineral water)            0.019997   
48      (french fries, spaghetti)  (mineral water)            0.027596   
49  (french fries, mineral water)      (spaghetti)            0.033729   

    consequent support   support  confidence      lift  leverage  conviction  
29            0.238368  0.059725    0.343032  1.439085  0.018223    1.159314  
5             0.238368  0.052660    0.321400  1.348332  0.013604    1.122357  
21            0.238368  0.047994    0.370370  1.553774  0.017105    1.209650  
15            0.238368  0.040928    0.416554  1.747522  0.017507    1.305401  
16            0.174110  0.039195    0.398915  2.291162  0.022088    1.373997  
..                 ...       ...         ...       ...       ...         ...  
60            0.174110  0.010265    0.371981  2.136468  0.005460    1.315071  
34            0.174110  0.010265    0.364929  2.095966  0.005368    1.300468  
45            0.238368  0.010132    0.506667  2.125563  0.005365    1.543848  
48            0.238368  0.010132    0.367150  1.540263  0.003554    1.203494  
49            0.174110  0.010132    0.300395  1.725318  0.004259    1.180509  

[63 rows x 9 columns]

## 4. Outras questões
Essa questão não possui correção automática.

### 4.a. Liste 5 cenários hipotéticos que você vê utilidade de se analisar regras de associação (pense em algum problema real, algum problema de pesquisa ou desenvolvimento que já atuou ou atua,...)

**RESPOSTA 4.a.:** 

1. Quase todas as lojas online se beneficiariam disso (roupas, livros, aparelhos eletrônicos) - Mesmo princípio que há na organização de supermercado, mas neste caso, sugestões ao estar visualizando um produto.
2. Métodos de estudos: Existem muitos fatores que influênciam a vida do estudante. Conseguir estudar e ter boas notas é extremante relativo. Talvez uma associação com diversos métodos de estudos, padrões comportamentais poderia ajudar a identificar algo relevante.
3. Biblioteca de jogos: Com base nos jogos que o cliente costuma comprar, associar as categorias dos mesmos, como rpg, fps etc. Com base nisso, promover promoções, mudanças no layout da aplicação para favorecer seu gosto e melhorar o acerto na visualização de lançamentos e prováveis compras.
4. Redes sociais: São muitos os conteúdos presentes nas redes sociais, de diversas àreas, ramificações. Tanto o conteúdo, quanto o usuários possuem muitas  características que o descrevem. Neste caso também poderiamos usar associações para fazer recomendação de posts ou produtos.
5. Manejo de solo: Em um caso hipotético, tendo todos os tipos e combinações feitas de plantações, montar um solo associando as melhores plantas combinadas juntas que propiciam uma boa colheita.

### 4.b. Considerando os resultados das regras, faça uma análise da regra (pancakes, mineral water)-> (spaghetti) em termos das métricas de "antecedent support",	"consequent support", "support",	"confidence" e	"lift". O objetivo é verificar se você entendeu os conceitos dessas métricas. Por exemplo, explique o que significa para essa regra um valor de "antecedent support" de 0.033729. Faça isso para todas as métricas.

**RESPOSTA 4.b.:**

In [ ]:
cond1 = rules['antecedents'].apply(lambda x: 'pancakes' in x and 'mineral water' in x)
cond2 = rules['consequents'].apply(lambda x: 'spaghetti' in x)
rules.loc[cond1 & cond2]

antecedents  consequents  antecedent support  \
62  (pancakes, mineral water)  (spaghetti)            0.033729   

    consequent support   support  confidence      lift  leverage  conviction  
62             0.17411  0.011465    0.339921  1.952333  0.005593    1.251198

- **antecedent support** *(0.033729)*: Frequência relativa com que a combinação(itemset) {"pancakes", "mineral water"} aparecem nas transações, 3.37%.

- **consequent support** *(0.17411)*: Frequêcia relativa com que a combinação(itemset) {"spaghetti"} aparece nas transações, 17.41%.

- **support** *(0.17411)*: Frequência relativa com que a combinação(itemset) {"pancakes", "mineral water", "spaghetti"} aparecem nas transações, ou seja, freq(XUY)/|T|, 17.41%.

- **confidence** *(0.339921)*: Diz respeito a porcentagem de vezes que o itemset {"spaghetti"} aparece quando há o itemset {"pancakes", "mineral water"}, ou seja, quando o cliente leva panquecas e água, quantos deles também levam espaguete, a porcentagem disso é a confiança, no caso, 33.9%.

- **lift** *(1.952333)*: O lift demonstra o quanto a presença do itemset {"pancakes", "mineral water"} na transação aumenta a probabilidade de conter {"spaghetti"}. Um valor maior que 1 quer dizer que o cliente comprando panquecas e água aumentam a probabilidade dele também levar espaguete.
Menor que 1 quer dizer que a probabilidade diminúi. Neste caso aumenta em 1.95x.

### 4.c. Utilize a função auxiliar abaixo para calcular conjuntos de itens frequentes (utilize suporte mínimo no intervalo [0.05, 0.3]) variando os tamanhos. Qual padrão você observa?
Qual variável você acredita que afeta o desempenho desse método? Experimente e discorra sobre quando usar esse método.

In [ ]:
# Gera uma matriz 2D de strings aleatórias.
#
# Parâmetro dv (different values): Quantidade de strings aleatórias únicas que a matriz irá conter.
# Parâmetro vpi (values per instance): Quantidade de strings aleatórias que cada instância terá.
# Parâmetro n: Quantidade de instâncias que será retornada.
def generate_fictional_data(dv=50, vpi=10, n=2500000):
    instances = [''.join(random.choices(string.ascii_uppercase + string.digits, k=20)) for _i in range(dv)]
    fictional_data = [random.choices(instances, k=vpi) for _i in range(n)]
    return fictional_data

In [133]:
def generate_frequent_itemsets(data):
  transactions = transactions_oht.fit(data).transform(data)
  df_data = pd.DataFrame(transactions, columns=transactions_oht.columns_)
  return apriori(df_data, min_support=0.05, use_colnames=True)

In [135]:

frequent_itemsets_2500000 = generate_frequent_itemsets(generate_fictional_data(n=2500000))
frequent_itemsets_250000 = generate_frequent_itemsets(generate_fictional_data(n=250000))
frequent_itemsets_25000 = generate_frequent_itemsets(generate_fictional_data(n=25000))
frequent_itemsets_2500 = generate_frequent_itemsets(generate_fictional_data(n=2500))
frequent_itemsets_250 = generate_frequent_itemsets(generate_fictional_data(n=250))
frequent_itemsets_25 = generate_frequent_itemsets(generate_fictional_data(n=25))

print("2500000 Strings")
print(frequent_itemsets_2500000[frequent_itemsets_2500000['support']<=0.3]['support'].describe())
print("250000 Strings")
print(frequent_itemsets_250000[frequent_itemsets_250000['support']<=0.3]['support'].describe())
print("25000 Strings")
print(frequent_itemsets_25000[frequent_itemsets_25000['support']<=0.3]['support'].describe())
print("2500 Strings")
print(frequent_itemsets_2500[frequent_itemsets_2500['support']<=0.3]['support'].describe())
print("250 Strings")
print(frequent_itemsets_250[frequent_itemsets_250['support']<=0.3]['support'].describe())
print("25 Strings")
print(frequent_itemsets_25[frequent_itemsets_25['support']<=0.3]['support'].describe())

2500000 Strings
count    50.000000
mean      0.182921
std       0.000234
min       0.182442
25%       0.182793
50%       0.182915
75%       0.183080
max       0.183408
Name: support, dtype: float64
250000 Strings
count    50.000000
mean      0.182857
std       0.000895
min       0.180820
25%       0.182337
50%       0.182774
75%       0.183600
max       0.184472
Name: support, dtype: float64
25000 Strings
count    50.000000
mean      0.183042
std       0.002375
min       0.176960
25%       0.181600
50%       0.183120
75%       0.184630
max       0.187040
Name: support, dtype: float64
2500 Strings
count    50.000000
mean      0.183384
std       0.007610
min       0.167600
25%       0.177400
50%       0.184000
75%       0.189600
max       0.196800
Name: support, dtype: float64
250 Strings
count    111.000000
mean       0.112685
std        0.065744
min        0.052000
25%        0.052000
50%        0.060000
75%        0.180000
max        0.244000
Name: support, dtype: float64
25 Strings
c

**RESPOSTA 4.c.:**

As informações acima descrevem o comportamento do suporte conforme ajustamos o tamanho do array de strings aleatórias.

A geração de strings aleatórias tem um limite para o número de strings aleatórias únicas. Isso é observado quando elevamos o número de combinações aleatórias, as keys geradas têm suporte alto, o que significa que aparecem constantemente em diferentes "transações". O parametro dv, se ajustado deverá
diminuir esse suporte novamente para os valores altos, entretanto, demorará para ser executado

Com uma média com baixo desvio padrão, um número alto gera keys que aparecem 18% nos dados. Ou seja, baixo caráter aleatório de keys.

Ao diminuir a quantidade de keys, os valores começam a se tornar mais variáveis, e com desvio padrão alto